src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs

```diff
@@ -0,0 +1,589 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Text.Json
+{
+    internal static partial class JsonWriterHelper
+    {
+        public static readonly byte[] s_newLineUtf8 = Encoding.UTF8.GetBytes(Environment.NewLine);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Really?

Reply:
It makes a 5-10% difference when writing strings (at depth 1).

Reply:
This amount of code bloat is not worth it. 

I believe I can write you a real-world looking benchmark - that runs code to compute the string, not just calls WriteString in a loop - where this code bloating optimization is going to make things 2x slower.

Reply:
> Isn't this something that happens at JIT time only 

Right, this code bloating happens at native code generation. It matters because of processors have instructions caches. The code that is in the code cache runs at full speed, the code that is not in the code cache takes performance hit.

When you are running microbenchmarks, all your code typically fits into the L1 cache and so you are getting the good case. The L1 code caches are not that big (in the order of 10s kB). When you are running real workloads, the code rarely fits into the L1 code cache but you still want to fit in as much of it as reasonably possible. Inlining large method is bad because of it wastes space in the code cache. The inlining of the large methods would be ok only if the real workloads are likely to use just a single copy of the inlined code for other reasons. I do not think it would be the case with TryWriteIndentation.

In other words, you want to keep in mind how much of the code cache is going to be occupied by your code in real workloads.

BTW: This was a real problem that we have hit with certain configs of TechEmpower. Our logic was essentially the same as the logic of other environments, the whole thing just run slower and we were struggling with figuring out why. The problem was traced down to code size: The code for our logic did not fit into the cache and run slower because of code cache misses.

Reply:
> The inlining of the large methods would be ok only if the real workloads are likely to use just a single copy of the inlined code for other reasons. I do not think it would be the case with TryWriteIndentation.

That is a great point that I hadn't considered. Thanks for the clarification!